### PR TITLE
Add assessment of interactive cluster usage compared to UC compute limitations

### LINKED ISSUE #498: Add assessment capability to locate use of boto3 and s3fs aws direct access libraries that typically assume Instance Profile access.
While reviewing customer's jobs, I come across ETL framework that leverages S3 to store config files. During migration, these should be moved to Volumes and Posix file access.

Upgrading clusters to UC (shared) clusters will eliminate embedded instance profiles. Most boto3 implementations rely on the underlying instance profile to access S3 files (and other AWS services).

The assessment / crawler should Identify Jobs and Notebooks that import boto3, s3fs and other libraries indicative of direct use of instance profiles. Use of boto3, s3fs cluster libraries should be flagged.

### LINKED ISSUE #810: [FEATURE]: Assess consumption of interactive shared clusters and look for usage patterns incompatible with shared clusters.
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem statement

UCX Assessment scans compute and tables but does not scan interactive usage of shared clusters. 
Understanding usage patterns of individual clusters (and users) will lend insights as to the related upgrade tasks and plan for interactive clusters and interactive users.

There are scenarios where data scientists are using techniques not directly compatible with UC or scenarios where developers are testing code for frameworks that are not directly compatible with UC. This assessment may catch and summarize these cases.

The audit logs can provide a rich input of what is run on interactive clusters and by whom. Scan the audit log (in system table form) to find code patterns to assess potential issues with migrating interactive clusters to `Shared` UC enabled clusters.

Adding this capability will require that the Unity Catalog `system.access.audit` table is available so this assessor will have to run as an optional task.

Related issues and enhancements: 
https://github.com/databrickslabs/ucx/issues/498 

### Proposed Solution

Add an Interactive Cluster audit log (system table) Assessment Task (or dashboard).
The query runs fast enough to be an interactive dashboard. Creating an inventory table would allow for capture of point in time snapshot of the assessment.

I just happen to have a prototype of the sql:
```sql
-- Scan notebook command history for potential paper cut issues
-- https://docs.databricks.com/en/compute/access-mode-limitations.html#compute-access-mode-limitations
-- This query 'overcounts' the paper cuts that might be encountered. There are complex DBR interactions with DBR 11, 12, 13 and 14
with paper_cut_patterns(
select col1 as pattern, col2 as issue from values
    ('hive_metastore.', 'I00'),
    ('spark.catalog.', 'I01.1'),
    ('spark._jsparkSession.catalog','I01.2'),
    ('spark._jspark', 'I02.1'),
    ('spark._jvm','I02.2'),
    ('._jdf', 'I02.3'),
    ('._jcol','I02.4'),
    ('spark.read.format("jdbc")', 'I04'),
    ('dbutils.notebook.entry_point.getDbutils().notebook().getContext().toJson()','I05.1'),
    ('dbutils.notebook.entry_point.getDbutils().notebook().getContext()','I05.2'),
    ('spark.udf.registerJavaFunction','I06'),
    ('boto3', 'I07.1'),
    ('s3fs', 'I07.2'),
    ('from graphframes', 'I08'),
    ('pyspark.ml.', 'I09'),
    ('applyInPandas', 'I10.1'),
    ('mapInPandas', 'I10.2'),
    ('dbutils.fs.', 'I11'),
    ('dbutils.credentials.', 'I12') -- credential passthrough
),
sparkcontext (
    select explode(split("_jvm, _jvm.org.apache.log4j, emptyRDD, range, init_batched_serializer, parallelize, pickleFile, textFile, wholeTextFiles, binaryFiles, binaryRecords, sequenceFile, newAPIHadoopFile, newAPIHadoopRDD, hadoopFile, hadoopRDD, union, runJob, setSystemProperty, uiWebUrl, stop, setJobGroup, setLocalProperty, getConf",', ')) as pattern,
    'I03.1' as issue
    UNION ALL
    select explode(split("from pyspark.sql import SQLContext, import org.apache.spark.sql.SQLContext, spark.sparkContext ", ', ')) as pattern, 'I03.2' as issue
),
streaming (
    select explode(split('.trigger(continuous, kafka.sasl.client.callback.handler.class, kafka.sasl.login.callback.handler.class, kafka.sasl.login.class, kafka.partition.assignment.strategy, kafka.ssl.truststore.location, kafka.ssl.keystore.location, cleanSource, sourceArchiveDir, applyInPandasWithState, .format("socket"), StreamingQueryListener',', ')) pattern,
    'I30' as issue
),
paper_cuts(
    select pattern, issue from paper_cut_patterns
    UNION ALL
    select concat('sc.',pattern) as pattern, issue from sparkcontext
    UNION ALL
    select pattern, issue from streaming
),
iteractive_cluster_commands (
    select 
        a.request_params.notebookId as notebook_id, 
        a.request_params.clusterId as cluster_id, 
        a.user_identity.email,
        a.request_params.commandLanguage,
        a.request_params.commandText 
    from system.access.audit a
    join system.compute.clusters as c
        ON c.cluster_source != 'JOB'
        AND (c.tags.ResourceClass is null OR c.tags.ResourceClass != "SingleNode")
        AND a.action_name = 'runCommand' 
        AND a.request_params.clusterId = c.cluster_id
    WHERE
        a.event_date >= DATE_SUB(CURRENT_DATE(), 90)
),
python_matcher(
    select 
    p.issue, 
    a.notebook_id, 
    a.cluster_id, 
    a.email,
    a.commandLanguage,
    a.commandText 
from iteractive_cluster_commands a
join paper_cuts p
    ON a.commandLanguage = 'python'
    AND contains(a.commandText, p.pattern)
),
scala_matcher(
    select
        'I20 - scala/R' as issue,
        a.notebook_id, 
        a.cluster_id, 
        a.email, 
        a.commandText  
    from iteractive_cluster_commands a
    where a.commandLanguage in ('scala','r')
),
unions(
    SELECT issue, notebook_id, cluster_id, email, commandText from python_matcher
    UNION ALL
    SELECT issue, notebook_id, cluster_id, email, commandText from scala_matcher
)
SELECT issue, 
    count(distinct notebook_id) distinct_notebooks,
    count(distinct cluster_id) distinct_clusters,
    count(distinct email) distinct_users
FROM unions group by 1
order by 1
;
```

### Additional Context

Summary view, the Issues (assessment findings) would be linked to the `assessment.md` document for more details. A short Assessment finding description could be provided.

image

### PR DESCRIPTION
**Pull request by dmoore247**
## Changes
Add new 'interactive' assessment report that providing insights into commands that have run on interactive compute and findings of potential issues in converting the clusters to Shared Unity Catalog data security mode.
image
image
image

### Linked issues
Resolves #810 
Resolves #498 

### Functionality 

- [x] added relevant user documentation
- [ ] added new CLI command
- [ ] modified existing command: `databricks labs ucx ...`
- [ ] added a new workflow
- [ ] modified existing workflow: `...`
- [ ] added a new table
- [ ] modified existing table: `...`
- [x] added new dashboard

### Tests

- [x] manually tested
- [ ] added unit tests
- [ ] added integration tests
- [x] verified on staging environment (screenshot attached)

### CODE REVIEW

File: docs/assessment.md [hunk 1]
```diff
@@ -350,6 +351,518 @@ CREATE TABLE [IF NOT EXISTS] table_name
...
+
+[[back to top](#migration-assessment-report)]
+
+### AF301.2 - spark.catalog.x
+
+The `spark._jsparkSession.catalog` pattern was found. Commonly used functions in spark.catalog, such as tableExists, listTables, setDefault catalog are not allowed/whitelisted on shared clusters due to security reasons.
+
+[[back to top](#migration-assessment-report)]
+
+### AF302.x - Arbitrary Java
```
Comment (nkvuong) [inline docs/assessment.md:372]: should this be a level higher than the one below

File: docs/assessment.md [hunk 2]
```diff
@@ -350,6 +351,518 @@ CREATE TABLE [IF NOT EXISTS] table_name
...
+For accessing non-storage cloud services (e.g., AWS secrets manager, etc), use
+
+
+(AWS) Consider other ways to authenticate with boto3, e.g., by passing credentials from Databricks secrets directly to boto3 as a parameter, or loading them as environment variables. This page contains more information. Please note that unlike instance profiles, those methods do not provide short-lived credentials out of the box, and customers are responsible for rotating secrets according to their security needs.
+
+
+(Azure) Consider creating a Databricks secrets scope backed by Azure vault, and authenticate to external systems using secrets from that secrets scope
+
+
+(GCP) Some customers have used an external secrets store (like Hashicorp vault) and loaded credentials upon cluster startup into the cluster’s environment variables. Please note that those are usually not short-lived credentials, and customers are responsible for securing and rotation secrets according to their security needs.
```
Comment (nkvuong) [inline docs/assessment.md:446]: boto3 is only used for AWS, so this advice for Azure & GCP wouldn't really apply

File: docs/assessment.md [hunk 3]
```diff
@@ -350,6 +351,518 @@ CREATE TABLE [IF NOT EXISTS] table_name
...
+
+
+[[back to top](#migration-assessment-report)]
+
+### AF305.2 - boto3
```
Comment (nkvuong) [inline docs/assessment.md:451]:
```suggestion
### AF305.2 - s3fs
```

File: docs/assessment.md [hunk 4]
```diff
@@ -350,6 +351,518 @@ CREATE TABLE [IF NOT EXISTS] table_name
...
+
+### AF310.1 - credential passthrough
+
+The `dbutils.credentials.` is used for credential passthrough. This is not supported by Unity Catalog.
+
+[[back to top](#migration-assessment-report)]
+
+### AF311.1 - dbutils.fs
+
+The `dbutils.fs.` was found. DBUtils and other clients that directly read the data from cloud storage are not supported.
```
Comment (nkvuong) [inline docs/assessment.md:477]: dbutils.fs. still work with volumes though - so we should highlight that

File: docs/assessment.md [hunk 5]
```diff
@@ -350,6 +351,518 @@ CREATE TABLE [IF NOT EXISTS] table_name
...
+
+[[back to top](#migration-assessment-report)]
+
+### AF313.25 - SparkContext
+
+The `.wholeTextFiles` was found.
+
+[[back to top](#migration-assessment-report)]
+
+### AF314.0 - Distributed ML
```
Comment (nkvuong) [inline docs/assessment.md:677]: higher level?

**Review by nkvuong — APPROVED** [review]
a few minor comments, but overall very useful :)

File: docs/assessment.md (hunk 2 above)
Comment (dmoore247) [reply docs/assessment.md:446]: Good point. Thanks!

File: docs/assessment.md [hunk 6]
```diff
@@ -350,6 +351,518 @@ CREATE TABLE [IF NOT EXISTS] table_name
...
+
+### AF305.2 - boto3
+
+The `s3fs` library is used which provides posix type sematics for S3 access. s3fs is based on boto3 and has similar restrictions. The recommendation is to use EXTERNAL VOLUMES mapped to the fixed s3 storage location or Unity Catalog MANAGED VOLUMES.
+
+[[back to top](#migration-assessment-report)]
+
+### AF306.1 - dbutils...getContext
+
+The `dbutils.notebook.entry_point.getDbutils().notebook().getContext().toJson()` was found. This function may trigger a security exception in DBR 13.0 and above.
```
[the remark this answers is not in the document]
Comment (dmoore247) [reply docs/assessment.md:459]: Of two minds here, `.safeToJson()` isn't documented, it isn't great alternative. A python repl function is supposed to be released.

File: docs/assessment.md (hunk 4 above)
Comment (dmoore247) [reply docs/assessment.md:477]: Will add.

File: docs/assessment.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (dmoore247) [reply docs/assessment.md:459]: Adding .safeToJson() and more

**Comment by dmoore247** [comment]
@nfx Incorporated @nkvuong 's feedback. Ready to merge. Additional improvements and ehancements can be a fast follow PR.

File: docs/assessment.md [hunk 7]
```diff
@@ -350,6 +351,581 @@ CREATE TABLE [IF NOT EXISTS] table_name
...
+
+The `hive_metastore.` is used to refer to a 3 level namespace. Most customers will migrate hive_metastore to a workspace catalog (named based on the workspace name). The code should then map the `hive_metastore` to the appropriate catalog name.
+
+Easier solution is to define a default catalog for your session, job, cluster or workspace.
+
+Setting the Workspace default, use the admin UI or command line:
+```
+% databricks unity-catalog metastores assign --workspace-id 1234567890123456 \
+                                             --metastore-id 12345678-9999-9de3-3456-e789f0a12b34 \
+                                           --default-catalog-name my_catalog
```
Comment (nkvuong) [inline docs/assessment.md:379]: this is not used anymore, the default catalog must be set using `databricks settings update-default-workspace-namespace`

Comment (dmoore247) [reply docs/assessment.md]: @nkvuong Nice catch! Updated

File: src/databricks/labs/ucx/queries/assessment/interactive/01_0_compute_access_mode_limitation_summary.sql [hunk 8]
```diff
@@ -0,0 +1,59 @@
+-- viz type=table, name=Compute Access Mode Limitation Summary, columns=issue,link,distinct_workspaces,distinct_notebooks,distinct_clusters,distinct_users
```
Comment (nfx) [inline src/databricks/labs/ucx/queries/assessment/interactive/01_0_compute_access_mode_limitation_summary.sql:1]: this folder must not be in the `assessment` folder, as it'll break `assessment` job on no-uc workspaces.

File: docs/assessment.md [hunk 9]
```diff
@@ -350,6 +351,579 @@ CREATE TABLE [IF NOT EXISTS] table_name
 
 [[back to top](#migration-assessment-report)]
 
+## AF300 - AF399
```
Comment (nfx) [inline docs/assessment.md:354]: re-run the TOC plugin to re-generate the table of contents

**Review by nfx — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/queries/assessment/interactive/01_0_compute_access_mode_limitation_summary.sql (hunk 8 above)
Comment (dmoore247) [reply src/databricks/labs/ucx/queries/assessment/interactive/01_0_compute_access_mode_limitation_summary.sql:1]: The `assessment` job doesn't call the `interactive` dashboard as of today. So we're safe for now. Will look at moving it as a safety precaution in a follow on PR.

**Pull request merged**